### PR TITLE
TS-1702: adds error handling on apply pages

### PR DESCRIPTION
**Pull request by sam-drumm**
https://hackney.atlassian.net/browse/TS-1702?atlOrigin=eyJpIjoiNWU5OGVmZDE4YzM4NGYwN2IyZjQ0OWQyODJmYjU5ZGUiLCJwIjoiaiJ9

**What**
Ensure that promises are fulfilled before progressing on the apply/ pages

**Why**
This is an implementation of the original issue detailed https://github.com/LBHackney-IT/lbh-housing-register/pull/421

**How**
Where the action triggers middleware to save, the custom hook is implemented to keep track of the promise state.

Where the action triggers a thunk directly, error handling is added to the page and the thunk is adapted to return an error on unfulfilled promise.

### CODE REVIEW

File: cypress/pages/ethnicity.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class EthnicityPage {
+  static getEthnicityPage() {
+    const testId = 'test-ethnicity-questions-page';
+    return cy.get(`[data-testid="${testId}"]`);
+  }
+  static visit() {
+    cy.visit(`/apply/submit/ethnicity-questions`);
+  }
+  // static getEthnicityRadios() {
```
Comment (LBHTKarki) [inline cypress/pages/ethnicity.ts:9]: comments

File: cypress/pages/ethnicity.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  // }
+  // static getSaveButton() {
+  //   const testId = 'test-submit-form-button';
+  //   return cy.get(`[data-testid="${testId}"]`);
+  // }
+  static getErrorSummary() {
+    const testId = 'test-ethnicity-questions-error-summary';
+    return cy.get(`[data-testid="${testId}"]`);
+  }
+  // static getLoadingSpinner() {
```
Comment (LBHTKarki) [inline cypress/pages/ethnicity.ts:21]: comments

File: package.json [hunk 3]
```diff
@@ -14,7 +14,7 @@
     "test": "jest && tsc --noEmit",
     "test:watch": "jest --watch",
     "prepare": "husky",
-    "cypress:open": "cypress open --browser electron",
+    "cypress:open": "cypress open --browser chrome",
```
Comment (LBHTKarki) [inline package.json:17]: is this on purpose?

File: pages/apply/submit/declaration.tsx [hunk 4]
```diff
@@ -27,72 +31,108 @@ const Declaration = (): JSX.Element => {
...
         getDisqualificationReasonOption(reason)
       );
       const reason = reasonStrings.join(',');
+      // this is just a notification, so for now we don't need to handle errors
       dispatch(sendDisqualifyEmail({ application, reason }));
-      dispatch(disqualifyApplication(application.id!));
-      router.push('/apply/not-eligible');
+
+      try {
+        console.log('hello');
```
Comment (LBHTKarki) [inline pages/apply/submit/declaration.tsx:50]: console.log

File: pages/apply/submit/declaration.tsx [hunk 5]
```diff
@@ -27,72 +31,108 @@ const Declaration = (): JSX.Element => {
...
+        await dispatch(disqualifyApplication(application.id!)).unwrap();
```
Comment (LBHTKarki) [inline pages/apply/submit/declaration.tsx:51]: `.unwrap()` here probably doesn't do anything unless chained with `then` and `catch`. Should the `router.push` happen inside `then` and errors handled in `catch` should the dispatch result in rejected action?

File: pages/apply/submit/declaration.tsx [hunk 6]
```diff
@@ -27,72 +31,108 @@ const Declaration = (): JSX.Element => {
...
+        router.push('/apply/not-eligible');
+      } catch (error) {
+        setUserError(error as string);
```
Comment (LBHTKarki) [inline pages/apply/submit/declaration.tsx:54]: should `setLoading(false);` be here too?

File: pages/apply/submit/declaration.tsx [hunk 7]
```diff
@@ -27,72 +31,108 @@ const Declaration = (): JSX.Element => {
...
       if (medicalNeeds > 0) {
+        // these are just notifications, so we don't need to handle errors yet
         dispatch(sendMedicalNeed({ application, medicalNeeds }));
       }
-
-      dispatch(completeApplication(application));
-      dispatch(createEvidenceRequest(application));
-      router.push('/apply/confirmation');
+      try {
+        await dispatch(completeApplication(application)).unwrap();
```
Comment (LBHTKarki) [inline pages/apply/submit/declaration.tsx:70]: same as above, would be good to review the setup when `unwrap()` is used.

File: pages/apply/submit/declaration.tsx [hunk 8]
```diff
@@ -27,72 +31,108 @@ const Declaration = (): JSX.Element => {
...
+        await dispatch(createEvidenceRequest(application)).unwrap();
+        router.push('/apply/confirmation');
+      } catch (error) {
+        console.log('Error completing the application:', error);
```
Comment (LBHTKarki) [inline pages/apply/submit/declaration.tsx:74]: `setLoading(false);` here too?

File: pages/apply/submit/ethnicity-questions.tsx [hunk 9]
```diff
@@ -1,5 +1,5 @@
 import { FormikValues, getIn } from 'formik';
-import { useRouter } from 'next/router';
+// import { useRouter } from 'next/router';
```
Comment (LBHTKarki) [inline pages/apply/submit/ethnicity-questions.tsx:2]: comment

File: pages/apply/submit/declaration.tsx (hunk 5 above)
Comment (LBHTKarki) [reply pages/apply/submit/declaration.tsx:51]: looks like unwrap() will cause an error to be thrown in this case, so to say it wouldn't do anything was not correct

File: pages/apply/submit/declaration.tsx (hunk 6 above)
Comment (sam-drumm) [reply pages/apply/submit/declaration.tsx:54]: the finally should set the loading state back to false in the try or catch

File: package.json (hunk 3 above)
Comment (sam-drumm) [reply package.json:17]: Yeah - seems more sensible to be loading with chrome with the redux dev-tools.

**Review by LBHTKarki — APPROVED** [review]
👍

**Pull request merged**